cuqi/geometry/_geometry.py

```diff
@@ -34,6 +35,11 @@ def variables(self):
                 self.variables = self.dim
         return self._variables
 
+    @property
+    def fun_dim(self):
+        """The dimension of the function space."""
+        return self.dim
```

Comment:
I am wary of assuming the default function value dimension is equal to the parameters. Perhaps it would be better to infer this property from `par2fun` by default instead? 

One suggestion would be something like
``` Python
    @property
    def fun_shape(self):
        """ The shape of the geometry (function space). """
        if self._fun_shape is None:
            # Attempt to infer dimension
            try:
                self._fun_shape = self.par2fun(np.ones(self.dim)).shape
            except Exception:
                pass
        return self._fun_shape
    
    @property
    def fun_dim(self):
        """ The dimension of the geometry (function space). """
        if self.fun_shape is None: return None
        return reduce(operator.mul, self.fun_shape) # math.prod(self.shape) for Python 3.8+
```


In fact we need to think about the parameter and function space shape and dimensions. I think these are inconsistent in various geometries matching either the parameter space or function space. It would be very very good to fix this. For example, currently the main assumption is that the `shape` attribute represents the *function space* shape in Image2D. In just turns out that `dim` matches the parameter space for this case, but that may not always be the case.

I would suggest we rethink this as part of this PR. In particular we can use names like `shape_par` and `shape_fun`, and `dim_par` and `dim_fun` to match. Let is discuss ASAP

Reply:
Thank you @nabriis for your suggestion. This is a very good point! Yes I think it will be nice to address it. Lets discuss this sometime soon.

Reply:
This is great point, now we have 4 properties (fun_shape, fun_dim, par_shape, par_dim). `par_shape` need to be defined, `par_dim` is derived from `par_shape`. `fun_shape` is inferred if possible or defines, `fun_dim` is inferred from `fun_shape`.

Now calling `funvals` and setting `Samples.is_par` to False, is only allowed for geometries where the function space is represented by a 1D ndarray.

Reply:
Great!